lib/nixd/include/nixd/AST.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include "CallbackExpr.h"
+#include "Scheduler.h"
+
+#include "lspserver/Path.h"
+#include "lspserver/Protocol.h"
+
+#include <llvm/ADT/FunctionExtras.h>
+#include <llvm/ADT/StringRef.h>
+
+#include <nix/error.hh>
+#include <nix/installable-value.hh>
+#include <nix/installables.hh>
+#include <nix/nixexpr.hh>
+
+namespace nixd {
+
+class EvaluationTask : public Task {
+
+public:
+  using CallbackAction = llvm::unique_function<void(std::exception *)>;
+
+private:
+  CallbackAction Action;
+
+  nix::ref<nix::InstallableValue> IValue;
+
+  nix::ref<nix::EvalState> State;
+
+public:
+  EvaluationTask(std::string Name, CallbackAction Action,
+                 nix::ref<nix::InstallableValue> IValue,
+                 nix::ref<nix::EvalState> State)
+      : Task(std::move(Name)), Action(std::move(Action)), IValue(IValue),
+        State(State) {}
+
+  void run() noexcept override;
+};
+
+/// A Nix language AST wrapper that support language features for LSP.
+class NixAST {
+#ifndef NDEBUG
+  bool Evaluated = false;
+#endif
+  nix::Expr *Root;
+  std::map<const nix::Expr *, nix::Value> ValueMap;
+  std::map<const nix::Expr *, nix::Env> EnvMap;
+
+public:
+  /// Inject myself into nix cache.
+  void injectAST(nix::EvalState &State, lspserver::PathRef Path);
+
+  /// From fresh-parsed nix::Expr *AST root, e.g. State->parseExprFromString
+  NixAST(ASTContext &Cxt, nix::Expr *Root);
+
+  /// Get the evaluation result (fixed point) of the expression.
+  nix::Value getValue(nix::Expr *Expr) {
```

Comment:
What happens if `ValueMap[Expr]` doesn't exist? return an error? or just assert.

Reply:
nice catch! I think should **assert** here, because `getValue` is expected to be called with ASTNodes managed by the instance, e.g. returned by `NixAST::lookup`

Reply:
should we return a reference iiuc?